sapi/fpm/fpm/fpm_unix.c

```diff
@@ -392,6 +402,93 @@ static int fpm_unix_conf_wp(struct fpm_worker_pool_s *wp) /* {{{ */
 }
 /* }}} */
 
+#if HAVE_FPM_CPUAFFINITY
+static long fpm_cpumax(void)
+{
+	static long cpuid = LONG_MIN;
+	if (cpuid == LONG_MIN) {
+		cpu_set_t cset;
+#if defined(HAVE_SCHED_SETAFFINITY)
+		if (sched_getaffinity(0, sizeof(cset), &cset) == 0) {
+#elif defined(HAVE_CPUSET_SETAFFINITY)
+		if (cpuset_getaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1, sizeof(cset), &cset) == 0) {
+#endif
+			cpuid = CPU_COUNT(&cset);
+		} else {
+			cpuid = -1;
+		}
+	}
+
+	return cpuid;
+}
+
+static void fpm_cpuaffinity_init(cpu_set_t *c)
+{
+	CPU_ZERO(c);
+}
+
+static void fpm_cpuaffinity_add(cpu_set_t *c, int min, int max)
+{
+	int i;
+
+	for (i = min; i <= max; i ++) {
+		if (!CPU_ISSET(i, c)) {
+			CPU_SET(i, c);
+		}
+	}
+}
+
+static int fpm_cpuaffinity_set(cpu_set_t *c)
+{
+#if defined(HAVE_SCHED_SETAFFINITY)
+	return sched_setaffinity(0, sizeof(c), c);
+#elif defined(HAVE_CPUSET_SETAFFINITY)
+	return cpuset_setaffinity(CPU_LEVEL_WHICH, CPU_WHICH_PID, -1, sizeof(c), c);
+#endif
+}
+
+static int fpm_setcpuaffinity(char *cpu_list)
+{
+	char *token, *buf;
+    cpu_set_t c;
+	int r, cpumax, min, max;
+
+	r = -1;
+	cpumax = fpm_cpumax();
+
+	fpm_cpuaffinity_init(&c);
+	token = php_strtok_r(cpu_list, ",", &buf);
```

Comment:
The problem here is that `php_strtok_r` modifies the value so you are changing the configured value in the config which should not be done. You should copy the list to a new buffer (either on heap allocated or using alloca) before calling this.